lib/grandpa/vote_message.go

```diff
@@ -70,13 +61,129 @@ func (s *Service) receiveVoteMessages(ctx context.Context) {
 					"prevote count %d, precommit count %d, votes needed %d",
 				v, vm.Message.AuthorityID, vm.Round, vm.Message.Stage,
 				s.lenVotes(prevote), s.lenVotes(precommit), s.state.threshold()+1)
+
+			// when a given vote is validated we should check
+			// if we have reached the prevotes threshold
+			prevotesThreshold := s.lenVotes(prevote) - 1
+			if !prevotesThresholdReached && prevotesThreshold >= int(threshold) {
+				prevotesThresholdReached = true
+				determinePrecommitCh <- struct{}{}
```

Comment:
why not just close the channel?

Reply:
It is possible to close the channel here but I believe is not the right choice because moving the `close(ch)` here we should remove the `defer close(ch)` and in a case where we should exit the `receiveVoteMessages` method without the condition was reached then the channel will keep open.